File: 3542-minimum-operations-to-convert-all-elements-to-zero.py
class Solution:
    def minOperations(self, nums: list[int]) -> int:
        '''
        We will track the layers using a monotonic (non-decreasing) stack.
        We have a loop that goes through each num
            - In each iteration,
                - we recursively pop from stack while num < stack[-1]
                - if num != stack[-1]
                    - add 1 operation to result
                - push num onto stack
        '''
        stack = [0]
        res = 0
        for num in nums:
            while num < stack[-1]:
                stack.pop()
            if num != stack[-1]:
                res += 1
            stack.append(num)
            
        return res


sol = Solution()
# sol.minOperations([2, 0, 3, 4, 0, 5, 6])
# sol.minOperations([0, 3, 4, 0, 5, 0])
print(f'output: {sol.minOperations([0, 2])}, expected: 1')
print(f'output: {sol.minOperations([3, 1, 2, 1])}, expected: 3')
print(f'output: {sol.minOperations([1, 2, 1, 2, 1, 2])}, expected: 4')

File: 3542-minimum-operations-to-convert-all-elements-to-zero.java
import java.util.*;

class MinimumOperationsToConvertAllElementsToZero {
    public int minOperations(int[] nums) {
        Deque<Integer> stack = new ArrayDeque<>();
        stack.push(0);
        int res = 0;
        for (int num : nums) {
            while (num < stack.peek()) {
                stack.pop();
            }
            if (num != stack.peek()) {
                res++;
            }
            stack.push(num);
        }
        return res;
    }

    public static void main(String[] args) {
        MinimumOperationsToConvertAllElementsToZero sol = new MinimumOperationsToConvertAllElementsToZero();
        System.out.println("Output: " + sol.minOperations(new int[] { 0, 2 }) + ", expected: 1");
        System.out.println("Output: " + sol.minOperations(new int[] { 3, 1, 2, 1 }) + ", expected: 3");
        System.out.println("Output: " + sol.minOperations(new int[] { 1, 2, 1, 2, 1, 2 }) + ", expected: 4");
    }
}
